File: Regression/Polynomial Regression/polynomial_regression.R

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# # Splitting the dataset into the Training set and Test set
# # install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting the Linear Regression Model
lin_reg = lm(formula = Salary ~ . , 
             data = dataset)

# Fiitng the Polynomial Regression Mopdel
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
poly_reg = lm( formula = Salary ~ . ,
               data = dataset)

# Visualising The Linear Regression Model 
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') + 
  ggtitle('Linear Model') + 
  xlab('Level') + 
  ylab('Salary')

# Visualising The Polynomial Regression Model 
X_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = X_grid  , y = predict(poly_reg , newdata = data.frame(Level = X_grid ,
                                                                          Level2 = X_grid^2,
                                                                          Level3 = X_grid^3))),
            color = 'blue') +
  ggtitle('Polynomial Model') + 
  xlab('Level') + 
  ylab('Salary')

# Predicting Linear Regression Results
y_pred = predict(lin_reg, data.frame(Level = 6.5))

# Predicting Polynomial Regression Results
y_pred = predict(poly_reg, data.frame(Level = 6.5,
                                     Level2 = 6.5^2,
                                     Level3 = 6.5^3))

File: Regression/Polynomial Regression/polynomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Fri Aug 17 23:45:43 2018

@author: Ritwik
"""

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
"""from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting the Linear Reression Model to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X , y)

# Fitting the Polynomial Rgression Model
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 10)
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualising the Linear Regression Results
plt.scatter(X , y , color = 'red')
plt.plot(X , lin_reg.predict(X) , color = 'blue')
plt.title('Linear Regression')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# Visualising the Polynomial Regression Results
X_grid = np.arange(min(X), max(X), 0.001)
X_grid = X_grid.reshape((len(X_grid),1))
plt.scatter(X , y , color = 'red')
plt.plot(X_grid , lin_reg_2.predict(poly_reg.fit_transform(X_grid)) , color = 'blue')
plt.title('Polynomial Regression')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# Prediccting a New Result for Linear Regression Model
y_pred = lin_reg.predict(6.5)

# Prediccting a New Result for Linear Regression Model
y_pred = lin_reg_2.predict(poly_reg.fit_transform(6.5))